File: packages/quran-engine-go/batch6_test.go
package quranengine

// Batch 6: the 99 Names in depth, and the chains of transmission of the Ten Readings.
//
// A deliberate translation of packages/quran-engine-js/test/batch6.test.js: the same corpora, the
// same counts, the same shape checks, so a port that drifts fails here rather than in a consumer.

import (
	"strings"
	"testing"
)

func foldBatch6(text string) string {
	var b strings.Builder
	for _, r := range text {
		switch {
		case r >= 0x064B && r <= 0x065F, r == 0x0670, r >= 0x06D6 && r <= 0x06ED, r == 0x0640:
			continue
		case r == 0x0671, r == 0x0623, r == 0x0625:
			b.WriteRune(0x0627)
		case r == 0x0624:
			b.WriteRune(0x0648)
		case r == 0x0626:
			b.WriteRune(0x064A)
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

func TestNamesDepthCounts(t *testing.T) {
	e := loadTestEngine(t)
	names, themes, occurrences := e.NamesDepthCount()
	if names != 99 || themes != 9 || occurrences != 194 {
		t.Fatalf("got (%d, %d, %d), want (99, 9, 194)", names, themes, occurrences)
	}
	for i, name := range e.NamesInDepth() {
		if name.Number != i+1 {
			t.Fatalf("name %d is out of order (number %d)", i, name.Number)
		}
	}
}

func TestEveryNameHasRootThemeAndProse(t *testing.T) {
	e := loadTestEngine(t)
	ids := map[string]bool{}
	for _, theme := range e.NameThemes() {
		ids[theme.ID] = true
	}
	for _, name := range e.NamesInDepth() {
		if name.Root == "" {
			t.Fatalf("name %d has no root", name.Number)
		}
		if !ids[name.Theme] {
			t.Fatalf("name %d theme %q is not in the list", name.Number, name.Theme)
		}
		if len(name.Explanation) <= 40 || len(name.Living) <= 20 {
			t.Fatalf("name %d prose is too short", name.Number)
		}
	}
}

func TestNameThemesPartition(t *testing.T) {
	e := loadTestEngine(t)
	total := 0
	for _, theme := range e.NameThemes() {
		members := e.NamesByTheme(theme.ID)
		if len(members) == 0 {
			t.Fatalf("theme %q has no Names", theme.ID)
		}
		total += len(members)
	}
	if total != 99 {
		t.Fatalf("themes cover %d Names, want 99", total)
	}
}

func TestRootIsSpacedAndClosesUp(t *testing.T) {
	e := loadTestEngine(t)
	rahman, ok := e.NameInDepth(1)
	if !ok || rahman.Root != "ر ح م" || RootKey(rahman.Root) != "رحم" {
		t.Fatalf("name 1 root is %q", rahman.Root)
	}
	for _, spelling := range []string{"رحم", "ر ح م"} {
		got := e.NamesByRoot(spelling)
		if len(got) != 2 || got[0].Number != 1 || got[1].Number != 2 {
			t.Fatalf("byRoot(%q) = %v", spelling, got)
		}
	}
}

func TestPlacedOccurrencesAreRealTokens(t *testing.T) {
	e := loadTestEngine(t)
	placed, unplaced := 0, 0
	for _, name := range e.NamesInDepth() {
		for _, o := range name.Occurrences {
			ayah := e.Ayah(o.Surah, o.Ayah)
			if ayah == nil {
				t.Fatalf("name %d points at %d:%d, which is not an ayah", name.Number, o.Surah, o.Ayah)
			}
			if o.Token == nil {
				if o.Tokens != 0 {
					t.Fatalf("name %d unplaced but spans %d", name.Number, o.Tokens)
				}
				unplaced++
				continue
			}
			tokens := strings.Fields(ayah.TextArabic)
			if *o.Token >= len(tokens) {
				t.Fatalf("name %d token %d past the end of %d:%d", name.Number, *o.Token, o.Surah, o.Ayah)
			}
			if foldBatch6(tokens[*o.Token]) == "" {
				t.Fatalf("name %d token %d folds to nothing", name.Number, *o.Token)
			}
			placed++
		}
	}
	if placed != 184 || unplaced != 10 {
		t.Fatalf("placed=%d unplaced=%d, want 184 and 10", placed, unplaced)
	}
}

func TestBasmalahCarriesTwoNames(t *testing.T) {
	e := loadTestEngine(t)
	hits := e.NamesInAyah(1, 3)
	if len(hits) != 2 {
		t.Fatalf("1:3 carries %d Names, want 2", len(hits))
	}
	for i, hit := range hits {
		if hit.Name.Number != i+1 || hit.Occurrence.Token == nil || *hit.Occurrence.Token != i {
			t.Fatalf("hit %d is name %d at %v", i, hit.Name.Number, hit.Occurrence.Token)
		}
	}
}

func TestNamesSearch(t *testing.T) {
	e := loadTestEngine(t)
	found := false
	for _, n := range e.SearchNamesInDepth("رحم", 25) {
		if n.Number == 1 {
			found = true
		}
	}
	if !found {
		t.Fatal("searching the root رحم does not find Ar-Rahman")
	}
	if len(e.SearchNamesInDepth("", 25)) != 0 {
		t.Fatal("an empty query returns results")
	}
}

func TestIsnadCounts(t *testing.T) {
	e := loadTestEngine(t)
	imams, narrators, companions := e.IsnadCount()
	if imams != 10 || narrators != 20 || companions != 13 {
		t.Fatalf("got (%d, %d, %d), want (10, 20, 13)", imams, narrators, companions)
	}
	if _, ok := e.IsnadProphet(); !ok {
		t.Fatal("no Prophet at the head of the chains")
	}
}

func TestEveryRiwayahResolvesTwoApiece(t *testing.T) {
	e := loadTestEngine(t)
	per := map[string]int{}
	for _, tag := range e.IsnadNarratorKeys() {
		imam, ok := e.IsnadImamOf(tag)
		if !ok {
			t.Fatalf("%s resolves to no imam", tag)
		}
		per[imam]++
	}
	for _, imam := range e.IsnadImamKeys() {
		if per[imam] != 2 {
			t.Fatalf("%s has %d narrators, want 2", imam, per[imam])
		}
	}
}

func TestEveryImamReachesCompanions(t *testing.T) {
	e := loadTestEngine(t)
	for _, imam := range e.IsnadImamKeys() {
		chain, _ := e.IsnadImam(imam)
		// Abu Jafar WAS a Successor and read on Companions himself, so he has no teachers layer.
		if imam != "Abu Jafar" && len(chain.Teachers) == 0 {
			t.Fatalf("%s has no teachers", imam)
		}
		if len(chain.Companions) == 0 {
			t.Fatalf("%s reaches no Companion", imam)
		}
		for _, node := range chain.Teachers {
			if node.Role != "successor" || !strings.HasPrefix(node.Detail, "d.") {
				t.Fatalf("%s teacher %q is %q / %q", imam, node.Name, node.Role, node.Detail)
			}
		}
		for _, node := range chain.Companions {
			if node.Role != "companion" {
				t.Fatalf("%s companion %q is %q", imam, node.Name, node.Role)
			}
		}
	}
}

func TestChainRunsProphetToNarrator(t *testing.T) {
	e := loadTestEngine(t)
	for _, tag := range e.IsnadNarratorKeys() {
		layers := e.IsnadChain(tag)
		if len(layers) < 4 {
			t.Fatalf("%s chain is only %d layers", tag, len(layers))
		}
		if layers[0].Title != "THE PROPHET" || layers[1].Title != "THE COMPANIONS" {
			t.Fatalf("%s starts %q, %q", tag, layers[0].Title, layers[1].Title)
		}
		imam, narrator := -1, -1
		for i, l := range layers {
			if l.Title == "THE IMAM" {
				imam = i
			}
			if l.Title == "THE NARRATOR" {
				narrator = i
			}
		}
		if imam < 0 || narrator < 0 || imam > narrator {
			t.Fatalf("%s has imam at %d and narrator at %d", tag, imam, narrator)
		}
	}
}

func TestImamChainEndsAtTwoNarrators(t *testing.T) {
	e := loadTestEngine(t)
	for _, imam := range e.IsnadImamKeys() {
		layers := e.IsnadChain(imam)
		last := layers[len(layers)-1]
		if last.Title != "HIS TWO NARRATORS" || len(last.Nodes) != 2 {
			t.Fatalf("%s ends with %q (%d nodes)", imam, last.Title, len(last.Nodes))
		}
	}
}

func TestReadsDirectlyMatchesLinks(t *testing.T) {
	e := loadTestEngine(t)
	for _, tag := range e.IsnadNarratorKeys() {
		chain, _ := e.IsnadNarrator(tag)
		if e.IsnadReadsDirectly(tag) != (len(chain.Links) == 0) {
			t.Fatalf("%s: readsDirectly disagrees with its links", tag)
		}
		sentence := e.IsnadSentence(tag)
		if sentence == "" {
			t.Fatalf("%s has no sentence", tag)
		}
		if strings.Contains(sentence, "did not meet") != (len(chain.Links) > 0) {
			t.Fatalf("%s: the sentence disagrees with its links", tag)
		}
	}
}

func TestHafsAndQunbul(t *testing.T) {
	e := loadTestEngine(t)
	if imam, _ := e.IsnadImamOf("Hafs an Asim"); imam != "Asim" {
		t.Fatalf("Hafs resolves to %q", imam)
	}
	if !e.IsnadReadsDirectly("Hafs an Asim") {
		t.Fatal("Hafs should read on Asim himself")
	}
	if !strings.HasPrefix(e.IsnadSentence("Hafs an Asim"), "Hafs read on Asim himself") {
		t.Fatal("Hafs's sentence is wrong")
	}
	if e.IsnadReadsDirectly("Qunbul an Ibn Kathir") {
		t.Fatal("Qunbul did not meet Ibn Kathir")
	}
	chain, _ := e.IsnadNarrator("Qunbul an Ibn Kathir")
	if len(chain.Links) != 3 {
		t.Fatalf("Qunbul has %d links, want 3", len(chain.Links))
	}
}

func TestUnknownIsnadKey(t *testing.T) {
	e := loadTestEngine(t)
	if len(e.IsnadChain("Nobody an Nobody")) != 0 || e.IsnadSentence("Nobody an Nobody") != "" {
		t.Fatal("an unknown key answered something")
	}
	if _, ok := e.IsnadImamOf("no separator here"); ok {
		t.Fatal("a tag with no separator resolved to an imam")
	}
}

File: packages/quran-engine-js/test/batch6.test.js
// @ts-check
/**
 * Batch 6: the 99 Names in depth, and the chains of transmission of the Ten Readings.
 *
 * The counts are pinned against the app's own packs, so a corpus that reaches the engine
 * half-imported fails loudly rather than quietly answering less than it should. The chains are
 * checked for SHAPE as well as size: every riwayah must resolve to one of the ten imams, and
 * every chain must start at the Prophet and end at a narrator or his students.
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { loadFromDisk } from "../src/node.js";
import { rootKey } from "../src/namesDepth.js";

const engine = await loadFromDisk();

// ---- the Names in depth ----------------------------------------------------------

test("namesDepth: all 99 Names carry depth, under nine themes", () => {
  const { names, themes, occurrences } = engine.namesDepth.count();
  assert.equal(names, 99);
  assert.equal(themes, 9);
  assert.equal(occurrences, 194);
  // Every number 1..99 is present exactly once, and they are in order.
  assert.deepEqual(engine.namesDepth.all().map((n) => n.number),
                   Array.from({ length: 99 }, (_, i) => i + 1));
});

test("namesDepth: every Name has a root, a theme the theme list knows, and prose", () => {
  const ids = new Set(engine.namesDepth.themes().map((t) => t.id));
  for (const name of engine.namesDepth.all()) {
    assert.ok(name.root.length > 0, `name ${name.number} has no root`);
    assert.ok(ids.has(name.theme), `name ${name.number} theme ${name.theme} is not in the list`);
    assert.ok(name.explanation.length > 40, `name ${name.number} explanation is too short`);
    assert.ok(name.living.length > 20, `name ${name.number} living line is too short`);
  }
});

test("namesDepth: the themes partition the 99 with nothing left over", () => {
  let total = 0;
  for (const theme of engine.namesDepth.themes()) {
    const members = engine.namesDepth.byTheme(theme.id);
    assert.ok(members.length > 0, `theme ${theme.id} has no Names`);
    total += members.length;
  }
  assert.equal(total, 99);
});

test("namesDepth: a root is spaced in the corpus and closed up for morphology", () => {
  const rahman = engine.namesDepth.byNumber(1);
  assert.ok(rahman);
  assert.equal(rahman.root, "ر ح م");
  assert.equal(rootKey(rahman.root), "رحم");
  // Ar-Rahman and Ar-Raheem are the same root, and `byRoot` takes either spelling.
  assert.deepEqual(engine.namesDepth.byRoot("رحم").map((n) => n.number), [1, 2]);
  assert.deepEqual(engine.namesDepth.byRoot("ر ح م").map((n) => n.number), [1, 2]);
});

test("namesDepth: a placed occurrence really is that ayah's token", () => {
  // The Names were placed against the app's own Hafs text, folded. Comparing literally would
  // call a pause mark a mismatch, so fold both sides the way batch 5 does.
  const marks = /[ً-ٰٟۖ-ۭـ]/g;
  const fold = (s) => s.replace(marks, "")
    .replace(/ٱ/g, "ا").replace(/[أإ]/g, "ا")
    .replace(/ؤ/g, "و").replace(/ئ/g, "ي");
  let placed = 0;
  let unplaced = 0;
  for (const name of engine.namesDepth.all()) {
    for (const o of name.occurrences) {
      const ayah = engine.quran.ayah(o.surah, o.ayah);
      assert.ok(ayah, `name ${name.number} points at ${o.surah}:${o.ayah}, which is not an ayah`);
      if (o.token === null) {
        // An unplaced occurrence names no token and spans none.
        assert.equal(o.tokens, 0, `name ${name.number} unplaced but spans ${o.tokens}`);
        unplaced += 1;
        continue;
      }
      const tokens = (ayah.textArabic ?? "").split(/\s+/).filter(Boolean);
      assert.ok(o.token < tokens.length,
                `name ${name.number} token ${o.token} past the end of ${o.surah}:${o.ayah}`);
      assert.ok(fold(tokens[o.token]).length > 0);
      placed += 1;
    }
  }
  assert.equal(placed, 184);
  assert.equal(unplaced, 10);
});

test("namesDepth: the two Names of the Basmalah are both found in 1:3, in token order", () => {
  const hits = engine.namesDepth.inAyah(1, 3);
  assert.deepEqual(hits.map((h) => h.name.number), [1, 2]);
  assert.deepEqual(hits.map((h) => h.occurrence.token), [0, 1]);
});

test("namesDepth: a Name can be found by root, explanation or living line", () => {
  assert.ok(engine.namesDepth.search("رحم").some((n) => n.number === 1));
  const first = engine.namesDepth.byNumber(1);
  const word = first.living.split(/\s+/).find((w) => w.length > 6);
  assert.ok(engine.namesDepth.search(word).length > 0);
  assert.deepEqual(engine.namesDepth.search(""), []);
});

// ---- the chains of transmission --------------------------------------------------

test("isnad: ten imams, twenty narrators, thirteen Companions", () => {
  const { imams, narrators, companions } = engine.isnad.count();
  assert.equal(imams, 10);
  assert.equal(narrators, 20);
  assert.equal(companions, 13);
  assert.ok(engine.isnad.prophet());
});

test("isnad: every riwayah resolves to one of the ten imams, two apiece", () => {
  const perImam = new Map(engine.isnad.imamKeys().map((k) => [k, 0]));
  for (const tag of engine.isnad.narratorKeys()) {
    const imam = engine.isnad.imamOf(tag);
    assert.ok(imam, `${tag} resolves to no imam`);
    perImam.set(imam, perImam.get(imam) + 1);
  }
  // The Ten Readings are each carried by exactly two narrators: that is what makes them twenty.
  for (const [imam, n] of perImam) assert.equal(n, 2, `${imam} has ${n} narrators`);
});

test("isnad: every imam reaches Companions, through Successors or directly", () => {
  for (const imam of engine.isnad.imamKeys()) {
    const chain = engine.isnad.imam(imam);
    // Abu Jafar is the one imam with no Successors between him and the Companions: he WAS a
    // Successor, who read on Ibn Abbas and Abu Hurayrah themselves. So teachers may be empty;
    // reaching a Companion may not.
    if (imam !== "Abu Jafar") assert.ok(chain.teachers.length > 0, `${imam} has no teachers`);
    assert.ok(chain.companions.length > 0, `${imam} reaches no Companion`);
    for (const node of chain.teachers) {
      assert.equal(node.role, "successor");
      assert.match(node.detail, /^d\./, `${node.name} has no death year`);
    }
    for (const node of chain.companions) assert.equal(node.role, "companion");
  }
});

test("isnad: a chain runs from the Prophet down to the narrator", () => {
  for (const tag of engine.isnad.narratorKeys()) {
    const layers = engine.isnad.chain(tag);
    // At least the Prophet, the Companions, the imam and the narrator: Abu Jafar's two have no
    // Successors layer, which is why the floor is four and not five.
    assert.ok(layers.length >= 4, `${tag} chain is only ${layers.length} layers`);
    assert.equal(layers[0].title, "THE PROPHET");
    assert.equal(layers[1].title, "THE COMPANIONS");
    // The narrator himself is always in there, and the imam always above him.
    const titles = layers.map((l) => l.title);
    assert.ok(titles.includes("THE IMAM"));
    assert.ok(titles.includes("THE NARRATOR"));
    assert.ok(titles.indexOf("THE IMAM") < titles.indexOf("THE NARRATOR"));
  }
});

test("isnad: an imam's chain ends at his two narrators", () => {
  for (const imam of engine.isnad.imamKeys()) {
    const layers = engine.isnad.chain(imam);
    const last = layers[layers.length - 1];
    assert.equal(last.title, "HIS TWO NARRATORS");
    assert.equal(last.nodes.length, 2, `${imam} does not end at two narrators`);
  }
});

test("isnad: reading directly is exactly having no links between", () => {
  for (const tag of engine.isnad.narratorKeys()) {
    const chain = engine.isnad.narrator(tag);
    assert.equal(engine.isnad.readsDirectly(tag), chain.links.length === 0, tag);
    // The sentence says which of the two it is, in so many words.
    const sentence = engine.isnad.sentence(tag);
    assert.ok(sentence.length > 0, tag);
    assert.equal(sentence.includes("did not meet"), chain.links.length > 0, tag);
  }
});

test("isnad: Hafs read on Asim himself; Qunbul did not meet Ibn Kathir", () => {
  assert.equal(engine.isnad.imamOf("Hafs an Asim"), "Asim");
  assert.ok(engine.isnad.readsDirectly("Hafs an Asim"));
  assert.match(engine.isnad.sentence("Hafs an Asim"), /^Hafs read on Asim himself/);

  assert.equal(engine.isnad.imamOf("Qunbul an Ibn Kathir"), "Ibn Kathir");
  assert.ok(!engine.isnad.readsDirectly("Qunbul an Ibn Kathir"));
  assert.equal(engine.isnad.narrator("Qunbul an Ibn Kathir").links.length, 3);
});

test("isnad: an unknown key answers nothing rather than throwing", () => {
  assert.deepEqual(engine.isnad.chain("Nobody an Nobody"), []);
  assert.equal(engine.isnad.sentence("Nobody an Nobody"), "");
  assert.equal(engine.isnad.imamOf("no separator here"), null);
});
